### ISSUE TITLE
Running await/yield from on a Deferred on 3.5 spews PendingDeprecationWarnings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8773]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8773|
|Type|defect|
|Created|2016-08-20 07:56:25Z|
|Branch|https://github.com/twisted/twisted/tree/8773-pep-479madness|

This makes the benchmarks run really, really slowly!

<details><summary>Searchable metadata</summary>

```
trac-id__8773 8773
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8773_pep_479madness 8773-pep-479madness
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1471679785000459 1471679785000459
changetime__1473844695351866 1473844695351866
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

PEP that introduced this change: https://www.python.org/dev/peps/pep-0479/

PR: https://github.com/twisted/twisted/pull/511

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Trunk (with -Wall) (sys.stdout writing is the huge bottleneck here):

```
19.0 deferred_await/sec (95 deferred_await in 5 seconds)
9.0 deferred_yieldfrom/sec (45 deferred_yieldfrom in 5 seconds)
```

This branch (with -Wall): (no warnings)

```
262.8 deferred_await/sec (1314 deferred_await in 5 seconds)
214.6 deferred_yieldfrom/sec (1073 deferred_yieldfrom in 5 seconds)
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Without `-Wall`; trunk:

```
263.0 deferred_await/sec (1315 deferred_await in 5 seconds)
```
this branch:
```
387.8 deferred_await/sec (1939 deferred_await in 5 seconds)
```

so still noticeable without the stderr hit :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Adding this to Deferred:

```
    def __iter__(self):
        return self
    __await__ = __iter__

    def __send__(self, value=None):
        result = getattr(self, 'result', _NO_RESULT)
        if result is _NO_RESULT:
            return self
        raise StopIteration(result)
    __next__ = __send__
```

appears to be nearly performance-equivalent, and a lot less ugly than resorting to `eval`.

Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

!^ is now implemented in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @hawkowl|
|-|-|

I think this might be cheating a little bit, since I did actually _type_ the code in question.  But I'm going to review it anyway, since it seems straightforwardly correct and the _commit_ belongs to `hawkowl` anyway.

Merge away!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset e38cc25a67747899c6984d6ebaa8d3d134799415
```
#!CommitTicketReference repository="" revision="e38cc25a67747899c6984d6ebaa8d3d134799415"
Merge 8773-pep-479madness: Make async/await + yield from on Python 3.5 faster

Author: hawkowl
Reviewer: glyph
Fixes: #8773
```
